### ISSUE TITLE
SourceSink Analysis and Reachability Analysis from Custom Entry Point

### ISSUE THREAD

**Issue by lkna** [arguslab/Argus-SAF#23]
This might have been asked on another issue, but I would like to perform the following task with Amandroid:

1. Given a URL (either in variable, or directly used as a parameter of a function), I want to know which Android API uses this URL. I read this documentation of [taint-analysis](http://pag.arguslab.org/argus-saf#tutorial-taint), but still confused which build-in managers I should use to achieve this goal or whether I should use custom source and sink manager.

2. After I found the API which uses the URL, I plan to make a custom entry point and do reachability analysis from that point.

I tried to do the 2nd point first, and here is what i tried so far:

val fileUri = FileUtil.toUri(getClass.getResource("/random/ie.errisbeo.app-1.apk").getPath)
val outputUri = FileUtil.toUri(getClass.getResource("/random").getPath)

val reporter = new DefaultReporter
val yard = new ApkYard(reporter)
val layout = DecompileLayout(outputUri)
val strategy = DecompileStrategy(new DefaultLibraryAPISummary(AndroidGlobalConfig.settings.third_party_lib_file), layout)
val settings = DecompilerSettings(debugMode = false, forceDelete = false, strategy, reporter)
val apk = yard.loadApk(fileUri, settings, collectInfo = true)

val clazz = apk.getClazz(new JawaType("org.osmdroid.contributor.GpxToPHPUploader$1")).get
val method: JawaMethod = clazz.getDeclaredMethodByName("run").get

val cfg = JawaAlirInfoProvider.getCfg(method)
val rda = JawaAlirInfoProvider.getRda(method,cfg)
val idfg = InterproceduralSuperSpark(apk,clazz.getDeclaredMethods.map(_.getSignature))

I don't know what to do from this point on. And another thing is that there is a possibility that the method resides on the third party library.

I also tried doing reachability analysis using AndroidReachingFactsAnalysis. But Amandroid did not generate environment for the 3rd party library. Thus I tried to generate the environment by calling AppInfoCollector.generateAsyncTask(apk,new JawaType("org.osmdroid.contributor.GpxToPHPUploader$1")), but I didn't appear in the apk.model.getComponents afterwards.

Thank you

**Comment by fgwei** [comment]
Maybe you can take a look at this example: https://github.com/arguslab/Argus-SAF-playground/blob/master/src/main/scala/org/argus/play/random/RetriveStringFromSourceOrSink.scala . The basic idea is first build a map between URL instance to url string. Then do the taint analysis and then find which URL instance is used in that sink.

**Comment by lkna** [comment]
I see. I will try that. About the reachability analysis, 
1. Is it better to use InterproceduralSuperSpark or AndroidReachingFactsAnalysis?
2. I want to confirm this, after I made the IDFG with AndroidReachingFactsAnalysis, I just have to call idfg.icfg.getCallGraph.getReachableMethods(clazz.getDeclaredMethods.map(_.getSignature)) to confirm that my method is in the IDFG, Am I doing this right?

Thank you for your kind reply

**Comment by fgwei** [comment]
1. Depends on your purpose, I recommend using AndroidReachingFactsAnalysis over InterProceduralSuperSpark, because InterProceduralSuperSpark is good at doing reachability analysis but not taint analysis.
2. This only tells you your method is reachable, you have to query the parameter value for that method call to confirm your targeted URL is used.

**Comment by lkna** [comment]
I see, my approach was to locate the sink method that use the URL, and then check whether that particular method is reachable or not. One last thing I want to confirm:
If i use InterProceduralSuperSpark the IDFG is by using your defined method as entry point
```
val clazz = apk.getClazz(new JawaType("org.osmdroid.contributor.GpxToPHPUploader$1")).get
val method: JawaMethod = clazz.getDeclaredMethodByName("run").get
val idfg = InterProceduralSuperSpark(apk, methods.map(_.getSignature))
```
while in AndroidReachingFactsAnalysis, you iterate from every entry point of the app and make IDFG out of it
```
val components = apk.model.getComponents
    components foreach{
      component =>
        apk.model.getEnvMap.get(component) match {
          case Some((esig, _)) =>
            println(esig)
            val ep = apk.getMethod(esig).get
            implicit val factory = new RFAFactFactory
            val initialfacts = AndroidRFAConfig.getInitialFactsForMainEnvironment(ep)
            val idfg = AndroidReachingFactsAnalysis(apk, ep, initialfacts, new ClassLoadManager, new Context(apk.nameUri), timeout = None)
           idfg.icfg.getCallGraph.getReachableMethods(clazz.getDeclaredMethods.map(_.getSignature))
        }
    }
```

**Comment by fgwei** [comment]
You are right. Of cause you can use AndroidReachingFactsAnalysis to start from your defined entry point as well. But in this case you have to prepare your own initialfacts.

**Comment by lkna** [comment]
Got it. It seems using AndroidReachingFactsAnalysis takes more time than using InterProceduralSuperSpark. Thank you for your response, that's so helpful.

**Comment by lkna** [comment]
Ah sorry I forgot to ask this, what is the different between AndroidReachingFactsAnalysis and Reaching Definition Analysis?

**Comment by lkna** [comment]
Another question about the source sink analysis. I used the AndroidUrlCollector to get all the URL and it seems there URL which was found by the AndroidUrlCollector but was not resolved as URL instance (such as resolved as @kind object).

Then, following your idea:
>The basic idea is first build a map between URL instance to url string. Then do the taint analysis and then find which URL instance is used in that sink.

1. I cannot map the URL string and URL instance since it is not resolved as URL in the jawa source code
2. My goal is to know the sink API method that use this URL, then what kind of taint analysis is suitable for my goal?

Thank you

**Comment by fgwei** [comment]
For your first point, can you tell me more detail of what you mean? or you can show me your example.

Second point, if you successfully build the map, then you can use the default taint analysis to achieve your goal.

**Comment by lkna** [comment]
Here is the screenshot of the jawa IR 
![screen shot 2017-10-24 at 14 03 34](https://user-images.githubusercontent.com/8587122/31925650-2a75fc2e-b8c5-11e7-9f11-e54dc6b48483.png)

The URL string http://www.m4rc3lv.nl/pub/reisgids/reisgids.html?lat= is used as the parameter of the StringBuilder and it is resolved as @kind object;

Isn't it supposed to be resolved as Ljava/net/URL;.<init>:(Ljava/lang/String;)V, or my understanding is wrong?

As the default taint analysis, which module is that? can you give a little bit more hint/example

Thank you

**Comment by fgwei** [comment]
This example is more complex. Because it will do a lot of string operation to generate the final URL. 

The line you pointing is assign the baseurl to a variable v3. (kind Object just means this is an object operation.) Then this baseurl is used in multiple stirng operation to fill in the args. It reaches the WebView.loadUrl in the end. 

There are not any Ljava/net/URL;.:(Ljava/lang/String;)V in your code, of course there will no such thing in the jawa.

**Comment by lkna** [comment]
I see. Not getting the full URL is OK for me, as long as I can get the information of what kind of sink API uses the baseURL (in this example it is the WebView.loadURL). So the default taint analysis, is it AndroidDataDependentTaintAnalysis?

**Comment by fgwei** [comment]
Yes, you can use AndroidDataDependentTaintAnalysis. You can add "WebView.loadURL" as sink point, and mark your baseUrl string creation as source.

**Comment by lkna** [comment]
Great! Thank you, you are really helpful.

**Comment by lkna** [comment]
For the sink, I currently i mark the sink by checking the signature, e.g.: if(sig == new Signature("Landroid/webkit/WebView;.loadUrl:(Ljava/lang/String;)V"))

Since I want to check for another type of sink, where I can get the list of signature? or Are there any other way to do this?

Another thing I want to ask, is it possible to determine a sink of a source without specifying the sink? (for example, in my case I specify the baseURL string creation as source, and I have to mark the sink by checking the signature, what if I just want to know the sink point of this URL without caring whether it is webview or not)
